File: ghidra_scripts/AnalyzeD2LoopsAndAI.py
#Analyze Diablo II Animation and AI Loops
#
#This script detects and documents tight loops used for:
#1. Animation frame processing
#2. AI decision making
#3. Monster behavior updates
#4. Movement path calculation
#
#Diablo II uses several distinct loop patterns for performance-critical code.
#This script identifies them and suggests appropriate optimizations.
#
#Key patterns detected:
#- Frame update loops (dwGfxFrame, dwFrameRemain increment/decrement)
#- Direction iteration loops (8-direction or 16-direction)
#- Monster AI state machines
#- Skill execution loops
#- Item list iteration loops
#
#Output:
#- Console listing of detected loops
#- Loop characteristics and optimization hints
#- Confidence scores for each detection
#
#@author Ben Ethington
#@category Diablo 2
#@description Detects and documents animation frames, AI loops, and performance-critical code patterns
#@keybinding
#@menupath Diablo II.Analyze Loops and AI
#
"""
Supporting documentation for Loop and AI Analyzer
"""

import re

class LoopAnalyzer:
    def __init__(self):
        self.listing = currentProgram.getListing()
        self.func_mgr = currentProgram.getFunctionManager()
        self.monitor = monitor
        self.loops_found = []

    def log(self, msg):
        print("[LoopAnalyzer] " + msg)

    def analyze_function(self, func):
        """Analyze a single function for loops"""
        func_name = func.getName()
        entry = func.getEntryPoint()

        if entry is None:
            return

        # Scan function for loop patterns
        loops = self.detect_loops(func)

        if loops:
            for loop_info in loops:
                self.loops_found.append({
                    'function': func_name,
                    'address': entry.toString(),
                    'loop_type': loop_info['type'],
                    'loop_start': loop_info['start'].toString(),
                    'loop_end': loop_info['end'].toString(),
                    'characteristics': loop_info['characteristics'],
                    'confidence': loop_info['confidence']
                })

    def detect_loops(self, func):
        """Detect loops in a function"""
        loops = []
        visited = set()

        entry = func.getEntryPoint()
        if entry is None:
            return loops

        # Scan for backward branches (loop indicators)
        body = func.getBody()
        instr = self.listing.getInstructionAt(entry)
        branch_targets = {}

        while instr is not None and body.contains(instr.getAddress()):
            addr = instr.getAddress().getOffset()

            # Find all branches
            if instr.getMnemonicString().startswith('J'):
                # Conditional or unconditional jump
                refs = instr.getFlows()
                for flow_addr in refs:
                    target_offset = flow_addr.getOffset()

                    # Backward branch = loop
                    if target_offset < addr:
                        loop_type, characteristics = self.classify_loop(instr, target_offset)

                        if loop_type:
                            loops.append({
                                'type': loop_type,
                                'start': flow_addr,
                                'end': instr.getAddress(),
                                'characteristics': characteristics,
                                'confidence': 0.8
                            })

            instr = instr.getNext()

        return loops

    def classify_loop(self, branch_instr, loop_start):
        """Classify loop by its characteristics"""
        characteristics = []
        loop_type = "Generic"

        # Scan the loop for specific patterns
        current = branch_instr.getPrevious()
        scan_depth = 0

        while current is not None and scan_depth < 50:
            scan_depth += 1
            op_str = current.toString()

            # Animation frame update pattern
            if 'dwGfxFrame' in op_str or 'frame' in op_str.lower():
                characteristics.append("Animation Frame Update")
                loop_type = "AnimationFrame"

            # Frame remaining pattern
            if 'dwFrameRemain' in op_str or 'remain' in op_str.lower():
                characteristics.append("Frame Counter Decrement")
                if loop_type != "AnimationFrame":
                    loop_type = "FrameCounter"

            # Direction iteration (8 or 16 directions)
            if 'direction' in op_str.lower() or 'dir' in op_str.lower():
                characteristics.append("Direction Iteration")
                loop_type = "DirectionLoop"

            # Mode/state machine pattern
            if 'mode' in op_str.lower() or 'state' in op_str.lower():
                characteristics.append("State Machine")
                if loop_type == "Generic":
                    loop_type = "StateLoop"

            # Monster/AI pattern
            if 'monster' in op_str.lower() or 'ai' in op_str.lower():
                characteristics.append("Monster/AI")
                if loop_type == "Generic":
                    loop_type = "AILoop"

            # List iteration
            if 'Next' in op_str or 'pNext' in op_str:
                characteristics.append("List Iteration")
                if loop_type == "Generic":
                    loop_type = "ListLoop"

            # Counter pattern (CMP with constant)
            if 'CMP' in current.getMnemonicString():
                match = re.search(r'CMP.*?,\s*(\d+)', op_str)
                if match:
                    count = int(match.group(1))
                    if count in [8, 16]:
                        characteristics.append(f"{count}-Direction Loop")
                        loop_type = "DirectionLoop"
                    elif count in [4, 32, 64]:
                        characteristics.append(f"Fixed Count: {count}")

            current = current.getPrevious()

        return loop_type, characteristics

    def analyze_all_functions(self):
        """Scan all functions for loops"""
        func_count = 0
        total_loops = 0

        funcs = list(self.func_mgr.getFunctions(True))
        total = len(funcs)

        self.log(f"Analyzing {total} functions for loops...")

        for func in funcs:
            func_count += 1

            if func_count % 100 == 0:
                self.log(f"Progress: {func_count}/{total}")

            self.analyze_function(func)

        total_loops = len(self.loops_found)
        return func_count, total_loops

    def print_results(self):
        """Print loop detection results"""
        print("\n" + "="*80)
        print("DIABLO II LOOP AND AI ANALYSIS RESULTS")
        print("="*80 + "\n")

        if not self.loops_found:
            print("[*] No significant loops detected in current analysis scope.\n")
            return

        # Group by loop type
        loops_by_type = {}
        for loop in self.loops_found:
            loop_type = loop['loop_type']
            if loop_type not in loops_by_type:
                loops_by_type[loop_type] = []
            loops_by_type[loop_type].append(loop)

        # Print by type
        for loop_type in sorted(loops_by_type.keys()):
            loops = loops_by_type[loop_type]
            print(f"{loop_type}: {len(loops)} detected")
            print("-" * 80)

            for i, loop in enumerate(sorted(loops, key=lambda x: x['confidence'], reverse=True)[:15]):
                print(f"  Function: {loop['function']}")
                print(f"  Address:  {loop['address']}")
                print(f"  Loop:     {loop['loop_start']} → {loop['loop_end']}")
                print(f"  Type:     {loop['loop_type']}")
                print(f"  Confidence: {loop['confidence']:.0%}")
                print(f"  Characteristics:")
                for char in loop['characteristics']:
                    print(f"    • {char}")
                print()

            if len(loops) > 15:
                print(f"  ... and {len(loops) - 15} more\n")

        # Summary
        print("="*80)
        print(f"SUMMARY: {len(self.loops_found)} loops detected")
        print("="*80 + "\n")

        print("""
ANALYSIS NOTES:

Animation Frame Loops:
  - Process unit animation frames
  - Decrement dwFrameRemain counter
  - Often use tight inner loops for performance
  - Candidates for __d2call custom calling convention

Direction Loops:
  - Iterate over 8 or 16 directions
  - Used for damage spread, targeting, pathfinding
  - Usually fixed iteration count
  - Key optimization target

Monster/AI Loops:
  - Process monster AI decisions
  - Update monster mode/state
  - Move monsters toward targets
  - Often have complex conditional logic

List Iteration Loops:
  - Walk linked lists (monsters, items, etc.)
  - Check next pointer (pNext, pRoomNext, etc.)
  - Common in room processing
  - May benefit from batching optimizations

Performance Optimization Opportunities:
  1. Use __d2call convention for loops with EBX parameters
  2. Consider loop unrolling for 8/16-direction iteration
  3. Cache frequently accessed pointers
  4. Minimize function calls inside loop bodies
  5. Consider SIMD instructions for vector operations
""")

    def generate_ghidra_comments(self):
        """Generate Ghidra comments for loops (requires MCP bridge)"""
        print("\n[*] Generating Ghidra comments for loops...")

        comments = []
        for loop in self.loops_found:
            comment = f"Loop ({loop['loop_type']}): {', '.join(loop['characteristics'][:3])}"
            comments.append({
                'address': loop['loop_start'],
                'comment': comment
            })

        return comments


def main():
    """Main entry point"""
    if currentProgram is None:
        print("[✗] No program loaded in Ghidra!")
        return

    print("[*] Starting Diablo II loop analysis...")
    print(f"[*] Program: {currentProgram.getName()}")

    analyzer = LoopAnalyzer()
    func_count, loop_count = analyzer.analyze_all_functions()

    analyzer.print_results()

    print(f"[✓] Analysis complete!")
    print(f"    Total functions scanned: {func_count}")
    print(f"    Loops detected: {loop_count}")

    # Optional: Generate Ghidra comments
    if loop_count > 0:
        comments = analyzer.generate_ghidra_comments()
        print(f"    Comments to apply: {len(comments)}")
        print("\n[*] Use MCP bridge to apply comments:")
        print("    batch_set_comments(function_address, decompiler_comments=comments)")


if __name__ == "__main__":
    main()

File: ghidra_scripts/Analyze_D2LoopsAndAI.java
// D2 Loops And AI
//
// Detects and documents tight loops used for animation frame processing, AI decision making, monster behavior, and movement path calculation in Diablo II binaries.
//
// Usage: Run from Script Manager on any D2 binary.
// Output: Console listing of detected loops with characteristics and confidence scores.
//
// @author Ben Ethington
// @category Diablo 2.Analysis
// @description Detect animation loops and AI patterns in D2 code
// @menupath Diablo 2.Analysis.D2 Loops And AI

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;
import java.util.*;
import java.util.regex.*;

public class Analyze_D2LoopsAndAI extends GhidraScript {

    private static class LoopInfo {
        String type;
        Address start;
        Address end;
        List<String> characteristics;
        double confidence;
        String functionName;
        String functionAddress;
    }

    private Listing listing;
    private FunctionManager funcMgr;
    private List<LoopInfo> loopsFound = new ArrayList<>();

    private List<LoopInfo> detectLoops(Function func) {
        List<LoopInfo> loops = new ArrayList<>();
        Address entry = func.getEntryPoint();
        if (entry == null) return loops;

        AddressSetView body = func.getBody();
        Instruction instr = listing.getInstructionAt(entry);

        while (instr != null && body.contains(instr.getAddress())) {
            long addr = instr.getAddress().getOffset();
            if (instr.getMnemonicString().startsWith("J")) {
                Address[] flows = instr.getFlows();
                for (Address flowAddr : flows) {
                    long targetOffset = flowAddr.getOffset();
                    if (targetOffset < addr) {
                        String loopType = "Generic";
                        List<String> characteristics = new ArrayList<>();
                        // classify loop
                        Instruction current = instr.getPrevious();
                        int scanDepth = 0;
                        while (current != null && scanDepth < 50) {
                            scanDepth++;
                            String opStr = current.toString();
                            String opLower = opStr.toLowerCase();
                            if (opStr.contains("dwGfxFrame") || opLower.contains("frame")) {
                                characteristics.add("Animation Frame Update");
                                loopType = "AnimationFrame";
                            }
                            if (opStr.contains("dwFrameRemain") || opLower.contains("remain")) {
                                characteristics.add("Frame Counter Decrement");
                                if (!loopType.equals("AnimationFrame")) loopType = "FrameCounter";
                            }
                            if (opLower.contains("direction") || opLower.contains("dir")) {
                                characteristics.add("Direction Iteration");
                                loopType = "DirectionLoop";
                            }
                            if (opLower.contains("mode") || opLower.contains("state")) {
                                characteristics.add("State Machine");
                                if (loopType.equals("Generic")) loopType = "StateLoop";
                            }
                            if (opLower.contains("monster") || opLower.contains("ai")) {
                                characteristics.add("Monster/AI");
                                if (loopType.equals("Generic")) loopType = "AILoop";
                            }
                            if (opStr.contains("Next") || opStr.contains("pNext")) {
                                characteristics.add("List Iteration");
                                if (loopType.equals("Generic")) loopType = "ListLoop";
                            }
                            if (current.getMnemonicString().equals("CMP")) {
                                Matcher m = Pattern.compile("CMP.*?,\\s*(\\d+)").matcher(opStr);
                                if (m.find()) {
                                    int count = Integer.parseInt(m.group(1));
                                    if (count == 8 || count == 16) {
                                        characteristics.add(count + "-Direction Loop");
                                        loopType = "DirectionLoop";
                                    } else if (count == 4 || count == 32 || count == 64) {
                                        characteristics.add("Fixed Count: " + count);
                                    }
                                }
                            }
                            current = current.getPrevious();
                        }
                        LoopInfo li = new LoopInfo();
                        li.type = loopType;
                        li.start = flowAddr;
                        li.end = instr.getAddress();
                        li.characteristics = characteristics;
                        li.confidence = 0.8;
                        li.functionName = func.getName();
                        li.functionAddress = func.getEntryPoint().toString();
                        loops.add(li);
                    }
                }
            }
            instr = instr.getNext();
        }
        return loops;
    }

    private void printResults() {
        println("\n" + "=".repeat(80));
        println("DIABLO II LOOP AND AI ANALYSIS RESULTS");
        println("=".repeat(80) + "\n");

        if (loopsFound.isEmpty()) {
            println("[*] No significant loops detected in current analysis scope.\n");
            return;
        }

        Map<String, List<LoopInfo>> byType = new LinkedHashMap<>();
        for (LoopInfo loop : loopsFound) {
            byType.computeIfAbsent(loop.type, k -> new ArrayList<>()).add(loop);
        }

        for (Map.Entry<String, List<LoopInfo>> entry : new TreeMap<>(byType).entrySet()) {
            List<LoopInfo> loops = entry.getValue();
            println(entry.getKey() + ": " + loops.size() + " detected");
            println("-".repeat(80));
            loops.sort((a, b) -> Double.compare(b.confidence, a.confidence));
            int shown = Math.min(15, loops.size());
            for (int i = 0; i < shown; i++) {
                LoopInfo loop = loops.get(i);
                println("  Function: " + loop.functionName);
                println("  Address:  " + loop.functionAddress);
                println("  Loop:     " + loop.start + " -> " + loop.end);
                println("  Type:     " + loop.type);
                println(String.format("  Confidence: %.0f%%", loop.confidence * 100));
                println("  Characteristics:");
                for (String c : loop.characteristics) {
                    println("    - " + c);
                }
                println("");
            }
            if (loops.size() > 15) {
                println("  ... and " + (loops.size() - 15) + " more\n");
            }
        }

        println("=".repeat(80));
        println("SUMMARY: " + loopsFound.size() + " loops detected");
        println("=".repeat(80) + "\n");

        println("ANALYSIS NOTES:\n");
        println("Animation Frame Loops:");
        println("  - Process unit animation frames");
        println("  - Decrement dwFrameRemain counter");
        println("  - Often use tight inner loops for performance");
        println("  - Candidates for __d2call custom calling convention\n");
        println("Direction Loops:");
        println("  - Iterate over 8 or 16 directions");
        println("  - Used for damage spread, targeting, pathfinding");
        println("  - Usually fixed iteration count\n");
        println("Monster/AI Loops:");
        println("  - Process monster AI decisions");
        println("  - Update monster mode/state\n");
        println("List Iteration Loops:");
        println("  - Walk linked lists (monsters, items, etc.)");
        println("  - Check next pointer (pNext, pRoomNext, etc.)\n");
        println("Performance Optimization Opportunities:");
        println("  1. Use __d2call convention for loops with EBX parameters");
        println("  2. Consider loop unrolling for 8/16-direction iteration");
        println("  3. Cache frequently accessed pointers");
        println("  4. Minimize function calls inside loop bodies");
    }

    @Override
    public void run() throws Exception {
        if (currentProgram == null) {
            println("[X] No program loaded in Ghidra!");
            return;
        }

        listing = currentProgram.getListing();
        funcMgr = currentProgram.getFunctionManager();

        println("[*] Starting Diablo II loop analysis...");
        println("[*] Program: " + currentProgram.getName());

        int funcCount = 0;
        FunctionIterator funcs = funcMgr.getFunctions(true);
        int total = funcMgr.getFunctionCount();
        println("[*] Analyzing " + total + " functions for loops...");

        while (funcs.hasNext()) {
            monitor.checkCanceled();
            Function func = funcs.next();
            funcCount++;
            if (funcCount % 100 == 0) {
                println("Progress: " + funcCount + "/" + total);
            }
            List<LoopInfo> loops = detectLoops(func);
            loopsFound.addAll(loops);
        }

        printResults();

        println("[+] Analysis complete!");
        println("    Total functions scanned: " + funcCount);
        println("    Loops detected: " + loopsFound.size());

        if (!loopsFound.isEmpty()) {
            println("    Comments to apply: " + loopsFound.size());
            println("\n[*] Use MCP bridge to apply comments:");
            println("    batch_set_comments(function_address, decompiler_comments=comments)");
        }
    }
}
